File: src/main/java/edu/cornell/cs/apl/viaduct/security/solver/JoinEdge.java
package edu.cornell.cs.apl.viaduct.security.solver;

import edu.cornell.cs.apl.viaduct.util.Lattice;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;
import org.jgrapht.graph.DefaultEdge;

/** Applies meet with a constant before passing on the incoming value. */
class JoinEdge<A extends Lattice<A>> extends DefaultEdge implements DataFlowEdge<A> {
  private final A elem;

  JoinEdge(A elem) {
    this.elem = elem;
  }

  public A getJoinConstant() {
    return this.elem;
  }

  @Override
  public A propagate(A other) {
    return this.elem.join(other);
  }
}

File: compiler/src/main/kotlin/io/github/apl_cornell/viaduct/algebra/solver/JoinEdge.kt
package io.github.apl_cornell.viaduct.algebra.solver

import io.github.apl_cornell.viaduct.algebra.Lattice
import io.github.apl_cornell.viaduct.util.dataflow.DataFlowEdge
import org.jgrapht.graph.DefaultEdge

/** Joins the incoming value with a constant before passing it on. */
internal class JoinEdge<A : Lattice<A>>(private val constant: A) : DefaultEdge(), DataFlowEdge<A> {
    override fun propagate(input: A): A {
        return constant.join(input)
    }

    override fun toString(): String {
        return "$constant ∨ _"
    }
}
